K8s guide (#26)
jobrunr/website: content/en/guides/advanced/k8s-autoscaling.md

jobrunr-example-854f975c57-fgrlz     0/1     ContainerCreating   0          1s
jobrunr-example-854f975c57-n49bx     0/1     ContainerCreating   0          2s
jobrunr-example-854f975c57-t9wvw     0/1     ContainerCreating   0          1s
jobrunr-example-854f975c57-vbdjr     0/1     ContainerCreating   0          1s
postgres-f879f4c78-6xhpg             1/1     Running             0          121m
```
Great! We see there are 5 new pods being created. After around 1 minute, the servers should announce themselves and will be visible in the JobRunr dashboard.

<figure>
    <img src="/documentation/k8s-autoscaling-servers.gif" class="kg-image">
</figure>

Once it's time for our scheduled jobs to execute, they'll be queued and then processed. Now the workers' usage metric kicks in! As we have 40 jobs and only 20 total workers, we will see 100% workers' usage throughout the processing of all the jobs. After 30 seconds (poll interval), we can observe another pod being created to match our specified 80% workers' usage target.

If we wait 5 minutes and don't schedule any new jobs, our deployment will scale back to 0.

> ## Optional - secure your JobRunr dashboard

Reviewer: ```suggestion
## Autoscaling your secured JobRunr dashboard
```

<figure style="width: 100%; max-width: 100%; margin: 0">

```yaml
...
    - type: metrics-api
      metricType: Value
      metadata:
        targetValue: "300"                                    # max 5 min latency
        activationTargetValue: "0"                            # activate when jobs have > 0 latency
        format: "json"
        url: "http://jobrunr-service.default:8000/api/metrics/jobs/enqueued"
        valueLocation: "latency"
```
</figure>

This trigger is defining that we want to scale up if there are jobs that are waiting in queue for more than 5 minutes.  

> Beware of the artificial latency that might be introduced by using [mutexes]({{<ref "documentation/pro/mutexes/_index.md">}}) or [priority queues]({{<ref "documentation/pro/priority-queues/_index.md">}}). You can always specify the queue you are interested in by adding the `queue` parameter in the enqueued jobs metrics url, see [JobRunr Pro metrics api]({{<ref "#jobrunr-pro-metrics-api">}}) for more info.

> We have defined three triggers, and when any of these triggers meet the criteria, KEDA will start scaling. The Horizontal Pod Autoscaler (HPA), which drives the scaling, will calculate proposed replica counts for each trigger and use the highest one to scale the workload to.

Reviewer: Should we add this to the following subsection instead? And make it conditional because not all user will be configured 3 triggers at the same time?

Another approach, would be to add it to the `Testing the autoscaling` section and illustrate it?

> We have defined three triggers, and when any of these triggers meet the criteria, KEDA will start scaling. The Horizontal Pod Autoscaler (HPA), which drives the scaling, will calculate proposed replica counts for each trigger and use the highest one to scale the workload to.

Reviewer: ```suggestion
We have defined three triggers, and when any of these triggers meet the criteria, KEDA will start scaling. The Horizontal Pod Autoscaler (HPA), which drives the scaling, will process the proposed replica counts for each trigger and use the highest one to scale the workload to.
```

Should we add a link to HPA documentation? I don't know if we already did.

## Finished example
<!-- TODO: Change the github url -->
The example app and k8s deployment files that we created in this guide are available in our [github repository](https://github.com/dumnicki/example-k8s-scaling).

## Limitations and considerations
Pay an extra attention when setting `minReplicaCount` property to 0. When no pods are running, workers' usage will always be 0 so to scale from 0 this metric has to be used with at least one additional metric (e.g. latency of enqueued jobs + workers usage). No recurring jobs can be executed if no `BackgroundJobServers` are running. You should have always 1 `BackgroundJobServer` running in.  

The workers' usage is updated at every [background job server poll interval](documentation/configuration/spring#advanced-configuration). Therefore, setting a shorter poll interval in KEDA when using only workers' usage metric is not recommended.

> If you are planning to run jobs with long execution times, you might want to prevent Kubernetes from terminating your replicas that are still processing a long-running job. To do that, you can modify the grace period for your pods in Kubernetes by setting the `terminationGracePeriodSeconds` property in your Pod `spec`. You'll also have to configure the duration to wait before interrupting jobs in JobRunr configuration. Imagine your longest running job takes 1 hour to complete. Then your grace period should be greater than the execution time of that job, for example 1.5 hours. You can configure your deployment as follows:

Reviewer: ```suggestion
If you are planning to run jobs with long execution times, you might want to prevent Kubernetes from terminating your replicas that are still processing a long-running job. To do that, you can modify the graceful termination period for your pods in Kubernetes. This can be done by setting the `terminationGracePeriodSeconds` property in your Pod `spec`. You'll also have to configure the duration to wait before interrupting jobs in JobRunr. 

The termination period of both Kubernetes and JobRunr can be determined by the maximum expected runtime of your jobs. Imagine your longest running job takes 1 hour to complete. Then your graceful period should be greater than the execution time of that job, for example an 1 hour and 30 minutes.

You can configure your deployment as follows:
```

<figure style="width: 100%; max-width: 100%; margin: 0">

```yaml
apiVersion: apps/v1
kind: Deployment
...
spec:
    ...
    template:
        ...
        spec:
            ...
            terminationGracePeriodSeconds: 5400
```
</figure>

> Similar wait period would also have to be configured in JobRunr. There are multiple different options to [configure JobRunr](/documentation/configuration/). Since we are using the `jobrunr-pro-spring-boot-3-starter` dependency, we can add the following property to the `application.properties`:

Reviewer: ```suggestion
Similarly, a termination period would also have to be configured for JobRunr. There are multiple different options to [configure JobRunr](/documentation/configuration/). Since we are using the `jobrunr-pro-spring-boot-3-starter` dependency, we can add the following property to the `application.properties`:
```

<figure style="width: 100%; max-width: 100%; margin: 0">

```properties
org.jobrunr.background-job-server.interrupt-jobs-await-duration-on-stop=90m
```
</figure>

> > If you are curious about how Kubernetes terminates the Pods, you can follow [this link](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination).

Reviewer: ```suggestion
> There are two things to note. First, the Kubernetes termination period should match or be larger than the JobRunr one. Second, contrary to JobRunr, Kubernetes does not terminate before the end of the configured period. So, even though, JobRunr may have stopped after only 10 minutes, the Pod may remain in terminating for long time doing nothing. If you are curious about how Kubernetes terminates the Pods, you can follow [this link](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination).
```

## Conclusion
In this guide, we’ve learned how to set up autoscaling in Kubernetes using KEDA and use the JobRunr Pro metrics api to create the scaling triggers.

---

<!-- TODO: This API reference should be moved to documentation -->

> ## JobRunr Pro metrics api

Reviewer: ```suggestion
## JobRunr Pro metrics API
```

---
title: Autoscale your JobRunr application deployed on Kubernetes

> description: This guide will help you use Jobrunr Pro features to create a Kubernetes deployment of your application that autoscales based on JobRunr provided metrics.

Reviewer 2: ```suggestion
description: This guide will help you to create a Kubernetes deployment of your application that autoscales based on metrics provided by JobRunr Pro.
```

### Kubernetes deployment
To run your application in the cluster with the autoscaling capability, we will need to deploy several services:
- [KEDA](https://keda.sh/) for autoscaling functionality,
- any [database supported by JobRunr]({{<ref "documentation/installation/storage/_index.md">}}),
- your application running JobRunr.

First, deploy your JobRunr application. You will need to enable the dashboard to use the [JobRunr Pro metrics API](#jobrunr-pro-metrics-api) which is going to be used to trigger autoscaling. You can deploy the background job server together with the dashboard or separately but make sure that the metrics API is always available for KEDA to fetch the metrics. 
For the purposes of this guide, we've deployed a simple Spring Boot application that allows us to schedule mock jobs via the `/schedule-example-jobs` endpoint.

Our setup consists of 2 deployments:
- `web` - which exposes the JobRunr dashboard on port `8000` and an API to schedule jobs on port `8080`
- `workers` - which enables background job server for background task processing

> > The separation of the `web` and the `workers` is there to enable us to scale those two different parts separately. The dashboard deployment can be scaled independently to handle varying levels of user traffic without affecting the job processing capabilities. Similarly, the `BackgroundJobServer` can scale based on job queue metrics, ensuring optimal job processing efficiency.  This separation ensures that even if no `BackgroundJobServers` are currently running, scheduled jobs will still be processed as the autoscaling configuration will start a new `BackgroundJobServer` on demand.

Reviewer 2: I would not mention scaling the dashboard based on traffic. Most of the time, it is only engineers from inside the company so the load will be really low. Also, multiple dashboards hit the DB rather hard due to a query (the one that shows how many jobs are enqueued / scheduled / ... ) only used in the dashboard.

Reviewer: I would update 

`The dashboard deployment can be scaled independently to handle varying levels of user traffic without affecting the job processing capabilities.` 

to 

`The web deployment can be scaled independently to handle varying levels of user traffic without affecting the job processing capabilities.`

Then the example makes sense, as they may have integrated the JobRunr dashboard with an existing web service from which jobs are enqueued (or used for other kind of services).

Author: ```suggestion
> The separation of the `web` and the `workers` is there to enable us to scale the number of replicas responsible for processing jobs, while maintaining only one instance of the dashboard. The `BackgroundJobServer` can scale based on job queue metrics, ensuring optimal job processing efficiency. This separation ensures that even if no `BackgroundJobServers` are currently running, scheduled jobs will still be processed as the autoscaling configuration will start a new `BackgroundJobServer` on demand.
```

## Why you should use JobRunr metrics to autoscale
Using JobRunr metrics for autoscaling ensures that your application scales dynamically based on real-time job processing needs. Unlike traditional autoscaling methods that rely solely on CPU or memory usage, JobRunr metrics allow you to scale based on specific workload factors such as workers' usage, job queue latency, and scheduled job count.  
This approach leads to more efficient resource utilization, faster job processing, and improved application performance, ultimately resulting in cost savings and a better user experience.

## Prerequisites
- JobRunr Pro 8.0.0 or later
- You already know how to configure JobRunr
- Basic knowledge of Kubernetes
- A working Kubernetes cluster

> In this guide, we will be working locally using [Minikube](https://minikube.sigs.k8s.io/). If you use a different Kubernetes setup, some steps might differ, and you may skip any `minikube` commands.

## Setup

> You can follow this guide step-by-step to deploy your existing JobRunr application. If you want to learn how to include and use JobRunr in your project, follow [this guide]({{<ref "guides/intro/java-lambda.md">}}) first.  

Author: I think even better would be:
> You can follow this guide step-by-step to deploy your existing JobRunr application. If you're new to JobRunr, start by following [this guide]({{<ref "guides/intro/java-lambda.md">}}) to learn how to include and use it in your project.

> You can follow this guide step-by-step to deploy your existing JobRunr application. If you want to learn how to include and use JobRunr in your project, follow [this guide]({{<ref "guides/intro/java-lambda.md">}}) first.  

Reviewer: Like the general but still not would not use 
> start by following

It sounds like an order instead of an invitation.

We assume you already know how to deploy your application on Kubernetes.  
In case you prefer to get straight to the solution, you can view the code from this guide [here]({{<ref "#finished-example">}}).

### Kubernetes deployment
To run your application in the cluster with the autoscaling capability, we will need to deploy several services:
- [KEDA](https://keda.sh/) for autoscaling functionality,
- any [database supported by JobRunr]({{<ref "documentation/installation/storage/_index.md">}}),

> - your application running JobRunr dashboard and your application running JobRunr `BackgroundJobServer`.

Author: Currently the metrics api is available only if the dashboard is enabled. Background job server has to be enabled but not necessarily separately. Maybe I should explain it in the guide.

jobrunr-dashboard-74db64f488-spv5g   1/1     Running             0          121m
jobrunr-example-854f975c57-2pmds     0/1     ContainerCreating   0          1s
jobrunr-example-854f975c57-fgrlz     0/1     ContainerCreating   0          1s
jobrunr-example-854f975c57-n49bx     0/1     ContainerCreating   0          2s
jobrunr-example-854f975c57-t9wvw     0/1     ContainerCreating   0          1s
jobrunr-example-854f975c57-vbdjr     0/1     ContainerCreating   0          1s
postgres-f879f4c78-6xhpg             1/1     Running             0          121m
```
Great! We see there are 5 new pods being created.  
After around 1 minute, the servers should announce themselves and will be visible in the JobRunr dashboard.

Once it's time for our scheduled jobs to execute, they'll be queued and then processed. Now the workers' usage metric kicks in! As we have 40 jobs and only 20 total workers, we will see 100% workers' usage throughout the processing of all the jobs. After 30 seconds (poll interval), we can observe another pod being created to match our specified 80% workers' usage target.

If we wait 5 minutes and don't schedule any new jobs, our deployment will scale back to 0.

> ## Optional - secure your JobRunr dashboard

Reviewer: Nice section. But I don't find it really useful in this context. I'd simply mention that if they have enabled authentication, KEDA supports different kind of authentication schemes and link to the official docs. Another reason is I don't want to advertise basic auth.

  namespace: default
spec:
  secretTargetRef:
    - parameter: username
      name: jobrunr-dashboard-secret
      key: username
    - parameter: password
      name: jobrunr-dashboard-secret
      key: password
```
</figure>

This specifies an authentication details that can be used in our triggers as follows:
<figure style="width: 100%; max-width: 100%; margin: 0">

```yaml
...
  triggers:
    - type: metrics-api
      ...
      metadata:
        ...
        authMode: "basic"
      authenticationRef:
        name: jobrunr-dashboard-creds
...
```
</figure>

Add it to all 3 of our triggers. Now, the autoscaling should work with the secured JobRunr endpoints.

## Finished example
<!-- TODO: Change the github url -->
The example app and k8s deployment files that we created in this guide are available in our [github repository](https://github.com/dumnicki/example-k8s-scaling).

## Limitations and considerations
Pay an extra attention when setting `minReplicaCount` property to 0. When no pods are running, workers' usage will always be 0 so to scale from 0 this metric has to be used with at least one additional metric (e.g. latency of enqueued jobs + workers usage).  

> No recurring jobs can be executed if no `BackgroundJobServers` are running. We recommend having always 1 `BackgroundJobServer` running.  

Reviewer: Should be on the same line as the previous sentence. Also I believe this may be more than a recommendation but a requirement in some scenarios? I wonder if we need the sentence.

> The workers' usage is updated at every [background job server poll interval](documentation/configuration/spring#advanced-configuration). Therefore, setting a higher poll interval in KEDA when using only workers' usage metric is not recommended.

Reviewer: I don't totally agree with this statement. While it's true that we save the maxOccupiedWorkers every poll interval, KEDA still needs to retrieve all the background job stats from the DB just to compute the usage. We can avoid extra DB load and CPU load by increasing the poll interval, am I wrong?

## Conclusion
In this guide, we’ve learned how to set up autoscaling in Kubernetes using KEDA and use the JobRunr Pro metrics api to create the scaling triggers.

---

<!-- TODO: This API reference should be moved to documentation -->
## JobRunr Pro metrics api
JobRunr Pro allows us to access useful metrics that can be used for autoscaling. This is an overview of them:
### Get Enqueued Jobs Metrics
`GET /api/metrics/jobs/enqueued?queue=[string]`

**Query parameters**: 
- `queue` get metrics only for jobs from the queue with a specified name.

#### Response content example
<figure style="width: 100%; max-width: 100%; margin: 0">

```json
{
    "latency":500,
    "count":10
}
```
</figure>

> `latency` - time the oldest Job has spent in the queue in milliseconds.

Reviewer: Should we usage bullet points here too?

`count` - number of jobs in the queue

### Get Scheduled Jobs Metrics
`GET /api/metrics/jobs/scheduled?runInMax=[string]&queue=[string]` 

**Query parameters**: 
- `runInMax` ISO-8601 duration format, get metrics for jobs scheduled to run in specified time at the latest. By default, we only count jobs that are scheduled to run no later than 1 minute from now.
- `queue` get metrics only for jobs from the queue with a specified name.

#### Response content example
<figure style="width: 100%; max-width: 100%; margin: 0">

```json
{
    "count":10
}
```
</figure>

`count` - number of jobs scheduled to run in max `runInMax` time

### Get Workers Metrics
`GET /api/metrics/workers` 

#### Response content example
<figure style="width: 100%; max-width: 100%; margin: 0">

```json
{
    "usage":0.25,
    "totalWorkers":8,
    "occupiedWorkers":2
}
```
</figure>

> `usage` - percentage value of total usage of workers

Reviewer: Should we usage bullet points here too?
